### PR TITLE
Bug/782 ux error handling

### LINKED ISSUE #782: Address crash to login in UX
Right now when the application receives an error response from the server the user is sent back to the login screen. An error screen should be created and have the user sent there instead. Their session is likely fine and the error may be recoverable. 

* Have the error page display upgraded severity if user continues to run into errors.

To accomplish the following needs to happen

- We need to catch non 2xx status codes during a state change `resolve`. This is done with the `$stateChangeError` message. So a handler needs to be placed there. This code should live within the routerHelper.
- To avoid locking the application into an endless loop trying to route the user anywhere we also want to keep track of the times the handler has caught an error and abort if a loop is detected.
- If the error is recoverable, likely not in a loop, then it should offer a way back to the application.

See this wiki for information on ui-router state error handling : https://github.com/angular-ui/ui-router/wiki

### PR DESCRIPTION
**Pull request by OBCENEIKON**
Closes #782

### CODE REVIEW

**Comment by stackus** [comment]
We shouldn't be putting error handling into the Authorization service. That is a mixing of concerns.

File: src/client/app/components/authentication/session-service.factory.js [hunk 1]
```diff
@@ -23,6 +23,8 @@
       service.email = data.email;
       service.role = data.role;
       service.updatedAt = data.updated_at;
+      service.errorCount = data.errorCount;
```
Comment (stackus) [inline src/client/app/components/authentication/session-service.factory.js]: Should probably put these values into their own Service or Model.

**Comment by OBCENEIKON** [comment]
Removed error handling from the authorization service and into it's own.

Fixes #782

File: src/client/app/states/error/error.state.js [hunk 2]
```diff
@@ -2,6 +2,7 @@
   'use strict';
 
   angular.module('app.states')
+    .directive('backButton', historyBack)
```
Comment (stackus) [inline src/client/app/states/error/error.state.js]: A back button service is being built in another feature branch. You can wait for that to merge, or change your back button to a simple button that goes to the dashboard.

The PR for the back button should be merged very soon.

File: src/client/app/components/error/http-error.state.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        // HttpErrorService.lastError = null;
+        HttpErrorService.clearError();
+      }
+
+      jQuery('html, body').animate({scrollTop: 0}, 200);
+    });
+
+    $rootScope.$on('$stateNotFound', function(event) {
+      console.log('httpError::$stateNotFound');
+      event.preventDefault();
```
Comment (stackus) [inline src/client/app/components/error/http-error.state.js]: Wouldn't we want to have it go to the 404 page?

File: src/client/app/components/error/http-error.state.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        // HttpErrorService.errors.splice(HttpErrorService.errors.indexOf(HttpErrorService.lastError, 1));
+        // HttpErrorService.lastError = null;
+        HttpErrorService.clearError();
+      }
+
+      jQuery('html, body').animate({scrollTop: 0}, 200);
+    });
+
+    $rootScope.$on('$stateNotFound', function(event) {
+      console.log('httpError::$stateNotFound');
```
Comment (stackus) [inline src/client/app/components/error/http-error.state.js]: debugging

File: src/client/app/components/error/http-error.state.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      HttpErrorService.logError(error);
+
+      event.preventDefault();
+
+      console.log(HttpErrorService.errors);
+      $state.go('error', { error: error, errorCount: HttpErrorService.errors });
+    });
+
+    $rootScope.$on('$stateChangeSuccess', function(event, toState, toParams, fromState, fromParams) {
+      console.log('httpError::$stateChangeSuccess');
```
Comment (stackus) [inline src/client/app/components/error/http-error.state.js]: debugging

File: src/client/app/components/error/http-error.state.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+      // Increment the error counter for the status code
+      // var errorArray = (SessionService.errors === undefined) ? [] : SessionService.errors;
+      // errorArray[error.status] = (errorArray[error.status] === undefined) ? 1 : errorArray[error.status] + 1;
+      // SessionService.errors = errorArray;
+      // SessionService.lastError = error.status;
+      HttpErrorService.logError(error);
+
+      event.preventDefault();
+
+      console.log(HttpErrorService.errors);
```
Comment (stackus) [inline src/client/app/components/error/http-error.state.js]: debugging

File: src/client/app/components/error/http-error.state.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    $rootScope.$on('$stateChangeStart', function(toState, toParams) {
+      console.log('httpError::$stateChangeStart');
+    });
+
+    // catch any error in resolve in state
+    $rootScope.$on('$stateChangeError', function(event, toState, toParams, fromState, fromParams, error) {
+      console.log('httpError::$stateChangeError');
+
+      // Increment the error counter for the status code
+      // var errorArray = (SessionService.errors === undefined) ? [] : SessionService.errors;
```
Comment (stackus) [inline src/client/app/components/error/http-error.state.js]: Commented code

File: src/client/app/components/error/http-error.state.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    function activate() {
+    }
+
+    $rootScope.$on('$stateChangeStart', function(toState, toParams) {
+      console.log('httpError::$stateChangeStart');
+    });
+
+    // catch any error in resolve in state
+    $rootScope.$on('$stateChangeError', function(event, toState, toParams, fromState, fromParams, error) {
+      console.log('httpError::$stateChangeError');
```
Comment (stackus) [inline src/client/app/components/error/http-error.state.js]: debugging

File: src/client/app/components/error/http-error.state.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+  /** @ngInject */
+  function appRun($rootScope, $state, HttpErrorService, jQuery) {
+    activate();
+
+    function activate() {
+    }
+
+    $rootScope.$on('$stateChangeStart', function(toState, toParams) {
+      console.log('httpError::$stateChangeStart');
```
Comment (stackus) [inline src/client/app/components/error/http-error.state.js]: debugging

File: src/client/app/components/error/http-error-service.factory.js [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+      clearError: clearError,
+      errors: [],
+      lastError: null
+    };
+
+    return service;
+
+    function logError(error) {
+      var errorCode = error.status;
+      service.errors[errorCode] = typeof service.errors[errorCode] === 'undefined' ? 1 : service.errors[errorCode] + 1;
```
Comment (stackus) [inline src/client/app/components/error/http-error-service.factory.js]: Angular provids a shorthand method for checking for undefined. `angular.isUndefined(foobar)`

**Comment by chriskacerguis** [comment]
This is so out of date, I'm just going to close it.

**Pull request closed without merging**